File: api/src/main/java/net/blueshell/api/model/Address.java
// Address.java
package net.blueshell.api.model;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Index;
import jakarta.persistence.Table;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.ToString;
import net.blueshell.api.base.BaseModel;
import org.hibernate.annotations.SQLDelete;
import org.hibernate.annotations.SQLRestriction;

@Entity
@Table(
        name = "addresses",
        indexes = {
                @Index(name = "idx_addresses_deleted_at", columnList = "deleted_at"),
                @Index(name = "idx_addresses_city", columnList = "city"),
                @Index(name = "idx_addresses_zip_code", columnList = "zip_code")
        }
)
@SQLDelete(sql = "UPDATE addresses SET deleted_at = NOW(), version = version + 1 WHERE id = ? AND version = ?")
@SQLRestriction("deleted_at = '9999-12-31 23:59:59'")
@Data
@EqualsAndHashCode(onlyExplicitlyIncluded = true, callSuper = true)
@NoArgsConstructor
@ToString(onlyExplicitlyIncluded = true, callSuper = true)
public class Address extends BaseModel {
    @Column
    @ToString.Include
    private String country;

    @Column
    @ToString.Include
    private String city;

    @Column
    @ToString.Include
    private String street;

    @Column(name = "house_number")
    @ToString.Include
    private String houseNumber;

    @Column(name = "zip_code")
    @ToString.Include
    private String zipCode;
}

File: api/src/main/kotlin/net/blueshell/api/user/persistence/Address.kt
package net.blueshell.api.user.persistence

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.Index
import jakarta.persistence.Table
import net.blueshell.api.shared.model.AuditedAutoIdEntity
import org.hibernate.annotations.SQLDelete
import org.hibernate.annotations.SQLRestriction

@Entity
@Table(
    name = "addresses",
    indexes = [
        Index(name = "idx_addresses_deleted_at", columnList = "deleted_at"),
        Index(name = "idx_addresses_city", columnList = "city"),
        Index(name = "idx_addresses_zip_code", columnList = "zip_code")
    ]
)
@SQLDelete(sql = "UPDATE addresses SET deleted_at = NOW(), version = version + 1 WHERE id = ? AND version = ?")
@SQLRestriction("deleted_at = '9999-12-31 23:59:59'")
class Address : AuditedAutoIdEntity() {
    @Column
    var country: String? = null

    @Column
    var city: String? = null

    @Column
    var street: String? = null

    @Column(name = "house_number")
    var houseNumber: String? = null

    @Column(name = "zip_code")
    var zipCode: String? = null
}
